crates/sloth_vm/src/vm.rs

```diff
@@ -0,0 +1,599 @@
+            panic!("Pointer referenced nothing");
+        };
+
+        match &obj.typ {
+            ObjectType::Function(function) => {
+                // Add a callstack entry for the function
+                let offset = self.stack.top - (function.arity as usize);
+                self.call_stack.push(CallFrame::new(offset, function));
+            }
+            ObjectType::NativeFunction(function) => {
+                let mut args = Vec::with_capacity(function.arity as usize);
+                for _ in 0..function.arity {
+                    args.push(self.stack.pop());
+                }
+
+                let name = function.name;
+                let returns_value = function.returns_value;
+
+                let internal = function.function;
+                let result = internal(self, &args);
+
+                match result {
+                    Ok(value) => {
+                        if returns_value {
+                            self.stack.push(value);
+                        }
+                    }
+                    Err(error) => match error {
+                        native::Error::InvalidArgument => {
+                            panic!("Invalid argument provided to '{name}'");
+                        }
+                        native::Error::Unknown(msg) => {
+                            panic!("Native function '{name}' failed due to '{msg}'");
+                        }
+                    },
+                }
+            }
+            _ => panic!("Object was not a function"),
+        }
+    }
+
+    fn call_return(&mut self) {
+        let function = self.call_stack.peek().function();
+        let stack_offset = self.call_stack.peek().stack_offset;
+
+        let return_value = if function.returns_value {
+            Some(self.stack.pop())
+        } else {
+            None
+        };
+
+        self.stack.top = stack_offset;
+
+        if let Some(return_value) = return_value {
+            self.stack.push(return_value);
+        }
+
+        self.call_stack.pop();
+    }
+
+    fn unwind(&mut self) {
+        unimplemented!("Implement unwinding for error handling");
+    }
+
+    #[inline(always)]
+    fn read_u8(&mut self) -> u8 {
+        let frame = self.call_stack.peek_mut();
+        let function = frame.function();
+        let byte = function.chunk.code[frame.pointer];
+        frame.pointer += 1;
+        byte
+    }
+
+    #[inline(always)]
+    fn read_u16(&mut self) -> u16 {
+        let frame = self.call_stack.peek_mut();
+        let chunk = &frame.function().chunk;
+
+        let bytes = (chunk.code[frame.pointer], chunk.code[frame.pointer + 1]);
+
+        frame.pointer += 2;
+
+        ((bytes.0 as u16) << 8) + (bytes.1 as u16)
+    }
+
+    #[inline(always)]
+    pub fn objects(&self) -> &ObjectMap {
+        &self.objects
+    }
+
+    #[inline(always)]
+    pub fn objects_mut(&mut self) -> &mut ObjectMap {
+        &mut self.objects
+    }
+}
+
+#[cfg(test)]
```

Comment:
Good tests, It looks pretty good, looks like you remembered to run `cargo-fmt` this tme so I don't have to do it for you.